src/client/mod.rs

```diff
@@ -1,322 +1,498 @@
+    /// let results = client.graphql("query { Project { id name } }", Some("1234"), None,
+    ///     "Project").await;
+    /// # }
     /// ```
-    #[allow(clippy::needless_doctest_main)]
-    pub async fn graphql(&mut self, query: &str, input: Option<&Value>) -> Result<Value, Error> {
+    pub async fn graphql(
+        &mut self,
+        query: &str,
+        partition_key: Option<&str>,
+        input: Option<&Value>,
+        result_field: &str,
+    ) -> Result<Value, Error> {
+        trace!(
+            "Client::graphql called -- query: {} | partition_key: {:#?} | input: {:#?} | result_field: {}",
+            query,
+            partition_key,
+            input,
+            result_field
+        );
+
         // format request body
         let req_body = json!({
             "query": query.to_string(),
             "variables": {
+                "partitionKey": partition_key,
                 "input": input
             }
         });
 
-        // send request
-        let client = reqwest::Client::new();
-        let resp = client
-            .post(self.endpoint.as_str())
-            .json(&req_body)
-            .send()
-            .await
-            .map_err(|e| Error::new(ErrorKind::ClientRequestFailed, Some(Box::new(e))))?;
+        debug!("Client::graphql making request -- req_body: {}", req_body);
+        let mut body = match self {
+            Client::Http { endpoint } => {
+                let client = reqwest::Client::new();
+                let response = client
+                    .post(endpoint.as_str())
+                    .json(&req_body)
+                    .send()
+                    .await?;
+                response.json::<serde_json::Value>().await?
+            }
+            Client::Local { engine } => {
+                let engine = engine.clone();
+                let (tx, rx) = mpsc::channel();
+                thread::spawn(move || {
+                    let metadata: HashMap<String, String> = HashMap::new();
+                    let result = from_value::<GraphQLRequest>(req_body)
+                        .map_err(|e| e.into())
+                        .and_then(|req| engine.execute(&req, &metadata));
+                    let _ = tx.send(result);
+                })
+                .join()
+                .expect("Thread panicked");
+                rx.recv()??
+            }
+        };
+        debug!("Client::graphql -- response body: {:#?}", body);
 
-        // parse result
-        let body = resp
-            .json::<serde_json::Value>()
-            .await
-            .map_err(|_e| Error::new(ErrorKind::ClientReceivedInvalidJson, None))?;
-
-        // extract data from result
-        match body.get("data") {
-            None => Err(Error::new(
-                ErrorKind::ClientRequestUnexpectedPayload(body.to_owned()),
-                None,
-            )),
-            Some(data) => Ok(data.to_owned()),
-        }
+        body.as_object_mut()
```

Comment:
I assume the idea behind this block is to take a graphql output that looks like:

```json
{
  "data": {
    "Project": <PROJECT>
  }
}
```

And return the user just `<PROJECT>` without the extra two layers of indirection. I think that in most cases (certainly in all the CRUD cases) this is the desired behavior. However, GraphQL supports multiple queries in a single request. For example, the following query:

```
query {
  Project {
    id
  }
  User {
    id
  }
}
```

Would return the following output:

```json
{
  "data": {
    "Project": <PROJECT>,
    "User": <USER>
  }
}
```

And this approach will prevent the caller from taking advantage of that functionality. What would you think of moving the logic that strips out the `result_field` to the individual CRUD functions (i.e. `create_node`) and allow this function to return the entire data object?

Reply:
Good catch.  Will be resolved in the next push.